sunscreen_runtime/src/builder.rs

```diff
@@ -0,0 +1,648 @@
+/// [`prove`][GenericRuntime::prove] function.
+pub struct ProofBuilder<'r, 'p, T: marker::Zkp, B: ZkpBackend> {
+    runtime: &'r GenericRuntime<T, B>,
+    program: &'p CompiledZkpProgram,
+    private_inputs: Vec<ZkpProgramInput>,
+    public_inputs: Vec<ZkpProgramInput>,
+    constant_inputs: Vec<ZkpProgramInput>,
+}
+
+impl<'r, 'p, T: marker::Zkp, B: ZkpBackend> ProofBuilder<'r, 'p, T, B> {
+    /// Create a new `ProofBuilder`. It's typically more convenient to create a proof builder
+    /// via [`runtime.proof_builder()`][GenericRuntime::proof_builder].
+    pub fn new(runtime: &'r GenericRuntime<T, B>, program: &'p CompiledZkpProgram) -> Self
+    where
+        T: marker::Zkp,
+        B: ZkpBackend,
+    {
+        Self {
+            runtime,
+            program,
+            private_inputs: vec![],
+            public_inputs: vec![],
+            constant_inputs: vec![],
+        }
+    }
+
+    /// Add a constant input to the proof builder.
+    pub fn constant_input(mut self, input: impl Into<ZkpProgramInput>) -> Self {
+        self.constant_inputs.push(input.into());
+        self
+    }
+
+    /// Add multiple constant inputs to the proof builder.
+    pub fn constant_inputs<I>(mut self, inputs: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        ZkpProgramInput: From<T>,
+    {
+        self.constant_inputs
+            .extend(inputs.into_iter().map(ZkpProgramInput::from));
+        self
+    }
+
+    /// Add a public input to the proof builder.
+    pub fn public_input(mut self, input: impl Into<ZkpProgramInput>) -> Self {
+        self.public_inputs.push(input.into());
+        self
+    }
+
+    /// Add multiple public inputs to the proof builder.
+    pub fn public_inputs<I>(mut self, inputs: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        ZkpProgramInput: From<T>,
+    {
+        self.public_inputs
+            .extend(inputs.into_iter().map(ZkpProgramInput::from));
+        self
+    }
+
+    /// Add a private input to the proof builder.
+    pub fn private_input(mut self, input: impl Into<ZkpProgramInput>) -> Self {
+        self.private_inputs.push(input.into());
+        self
+    }
+
+    /// Add multiple private inputs to the proof builder.
+    pub fn private_inputs<I>(mut self, inputs: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+        ZkpProgramInput: From<T>,
+    {
+        self.private_inputs
+            .extend(inputs.into_iter().map(ZkpProgramInput::from));
+        self
+    }
+
+    /// Generate a proof; see [`runtime.prove()`][GenericRuntime::prove].
+    pub fn prove(self) -> Result<Proof> {
+        self.runtime.prove(
+            self.program,
+            self.private_inputs,
+            self.public_inputs,
+            self.constant_inputs,
+        )
+    }
+
+    /// Generate a proof with parameters for that proof system; see
+    /// [`runtime.prove_with_parameters()`][GenericRuntime::prove_with_parameters].
+    pub fn prove_with_parameters(
```

Comment:
Will people need to use this publically? If the moving around of shared generators and blinding factors is handled then they probably don't need this function exposed on the top level.

Reply:
I have no idea, this was added to the existing builders at some point, so I left it as is.

Reply:
I had added the prove with parameters as a stop gap to manually pass the parameters around but if we are doing it automatically I don't think the user needs to see this function.

Reply:
Ok I've removed them from the builder and removed a runtime `prove_with_parameters` that was unused. Marked the `verify_with_parameters` as `pub(crate)`.